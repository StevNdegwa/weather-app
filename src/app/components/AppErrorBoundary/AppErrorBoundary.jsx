import { Component } from "react";
import ErrorPage from "../ErrorPage";

class AppErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
  }

  static getDerivedStateFromError(error) {
    return { error };
  }

  render() {
    if (this.state.error) {
      return (
        <ErrorPage
          errorMessage={String(this.state.error) || "An unknown error occured"}
        />
      );
    }

    return this.props.children;
  }
}

export default AppErrorBoundary;
