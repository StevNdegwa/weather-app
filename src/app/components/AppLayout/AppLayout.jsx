import { AppLayoutWrapper, WeatherIconContainer } from "./style";
import WeatherIcon from "../WeatherIcon";

export default function AppLayout({ children }) {
  return (
    <AppLayoutWrapper>
      {children}
      <WeatherIconContainer>
        <WeatherIcon />
      </WeatherIconContainer>
    </AppLayoutWrapper>
  );
}
