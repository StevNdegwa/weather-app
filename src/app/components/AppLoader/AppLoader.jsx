import { Box } from "@mui/material";
import Loader from "react-loader-spinner";
import { AppLoaderWrapper } from "./styles";

export default function AppLoader() {
  return (
    <AppLoaderWrapper>
      <Box>
        <Loader type="ThreeDots" color="white" />
        <Box>In a moment . . .</Box>
      </Box>
    </AppLoaderWrapper>
  );
}
