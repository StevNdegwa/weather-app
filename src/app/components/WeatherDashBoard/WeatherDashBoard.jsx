import { useCallback, useMemo, useRef, useState } from "react";
import { Fade } from "@mui/material";
import PropTypes from "prop-types";
import TemperatureChart from "../TemperatureChart";
import DataLoadingIndicator from "../DataLoadingIndicator";
import WeatherDashboardHeader from "./WeatherDashboardHeader";
import WebDashBoardFooter from "./WebDashBoardFooter";
import WeatherDashBoardMain from "./WeatherDashBoardMain";
import { WeatherDashBoardWrapper, WeatherDashBoardContent } from "./styles";
import getDayWeatherAverage from "../../utils/getDayWeatherAverage";

export default function WeatherDashBoard({
  fadeIn,
  forecast = [],
  loadingData,
  reloadData,
}) {
  const [showingTempChart, setShowingTempChart] = useState(false);
  const sliderRef = useRef();
  const openTempChart = useCallback(
    () => setShowingTempChart(true),
    [setShowingTempChart]
  );
  const closeTempChart = useCallback(
    () => setShowingTempChart(false),
    [setShowingTempChart]
  );

  const slideToHome = useCallback(() => {
    sliderRef.current.goTo(0);
  }, []);

  const dailyAvgWeather = useMemo(() => {
    return getDayWeatherAverage(forecast);
  }, [forecast]);

  return (
    <Fade in={fadeIn}>
      <WeatherDashBoardWrapper>
        <TemperatureChart open={showingTempChart} closeModal={closeTempChart} />
        <DataLoadingIndicator loading={loadingData} />
        <WeatherDashBoardContent>
          <WeatherDashboardHeader reloadData={reloadData} />
          <WeatherDashBoardMain
            ref={sliderRef}
            forecastData={dailyAvgWeather}
            openTempChart={openTempChart}
          />
          <WebDashBoardFooter slideToHome={slideToHome} />
        </WeatherDashBoardContent>
      </WeatherDashBoardWrapper>
    </Fade>
  );
}

WeatherDashBoard.propTypes = {
  fadeIn: PropTypes.bool.isRequired,
  loadingData: PropTypes.bool.isRequired,
  reloadData: PropTypes.func.isRequired,
  error: PropTypes.object,
  forecast: PropTypes.arrayOf(
    PropTypes.shape({
      date: PropTypes.string,
      temperature: PropTypes.number,
    })
  ),
};
