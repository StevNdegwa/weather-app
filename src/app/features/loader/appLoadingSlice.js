import { createSlice } from "@reduxjs/toolkit";

const appLoadingSlice = createSlice({
  name: "appLoading",
  initialState: true,
  reducers: {
    appLoaded() {
      return false;
    }
  }
})

export const { appLoaded } = appLoadingSlice.actions;
export default appLoadingSlice.reducer;