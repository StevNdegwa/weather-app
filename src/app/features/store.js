import { configureStore } from "@reduxjs/toolkit";
import reducer from "./reducers";
import { weatherApiMiddleware } from "./weather/weatherForecastApi";

const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(weatherApiMiddleware),
})

export default store;