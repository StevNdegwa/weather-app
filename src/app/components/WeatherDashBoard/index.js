import WeatherDashBoard from "./WeatherDashBoard";
export default WeatherDashBoard;