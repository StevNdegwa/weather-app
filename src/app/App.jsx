import { GlobalStyles } from "@mui/material";
import { useSelector, useDispatch } from "react-redux";
import globalStyles from "./globalStyles";
import { useGetWeatherForecastQuery } from "./features/weather/weatherForecastApi";
import { appLoaded } from "./features/loader/appLoadingSlice";
import AppLoader from "./components/AppLoader";
import AppLayout from "./components/AppLayout";
import WeatherDashBoard from "./components/WeatherDashBoard";
import AppErrorBoundary from "./components/AppErrorBoundary";
import ErrorPage from "./components/ErrorPage";

export default function App() {
  const loadingApp = useSelector((state) => state.appLoading);
  const setLocation = useSelector((state) => state.setLocation);
  const dispatch = useDispatch();

  const { data, error, isLoading, isFetching, refetch } =
    useGetWeatherForecastQuery(setLocation.name);

  const showAppLoader = loadingApp && !error;
  const showDashBoard = !error && !loadingApp;

  if (data && loadingApp) {
    dispatch(appLoaded());
  }

  return (
    <AppLayout>
      <AppErrorBoundary>
        {Boolean(error) && <ErrorPage errorMessage={error.error} />}
        {showAppLoader && <AppLoader />}
        {showDashBoard && (
          <WeatherDashBoard
            fadeIn={!loadingApp}
            forecast={data ? data.forecast : []}
            loadingData={isFetching || isLoading}
            reloadData={refetch}
          />
        )}
      </AppErrorBoundary>
      <GlobalStyles styles={globalStyles} />
    </AppLayout>
  );
}
