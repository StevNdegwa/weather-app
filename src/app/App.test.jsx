import { mount } from "enzyme";
import fetchMock from "jest-fetch-mock";
import { Provider } from "react-redux";
import { GlobalStyles } from "@mui/material";
import { appLoaded } from "./features/loader/appLoadingSlice";
import store from "./features/store";
import AppLoader from "./components/AppLoader";
import AppLayout from "./components/AppLayout";
import WeatherDashBoard from "./components/WeatherDashBoard";
import AppErrorBoundary from "./components/AppErrorBoundary";

import App from "./App";

describe("test <App/> component", () => {
  let wrapper;

  beforeEach(() => {
    fetchMock.resetMocks();

    wrapper = mount(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });

  test("snapshot testing", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should render AppLayout component", () => {
    expect(wrapper.find(AppLayout)).toHaveLength(1);
  });

  test("should render all the components by default", () => {
    expect(wrapper.find(GlobalStyles)).toHaveLength(1);
    expect(wrapper.find(AppLoader)).toHaveLength(1);
    expect(wrapper.find(WeatherDashBoard)).toHaveLength(0);
    expect(wrapper.find(AppErrorBoundary)).toHaveLength(1);
  });

  test("should not render the application loader", () => {
    store.dispatch(appLoaded());
    wrapper.update();
    expect(wrapper.find(AppLoader)).toHaveLength(0);
    expect(wrapper.find(WeatherDashBoard)).toHaveLength(1);
  });

});
