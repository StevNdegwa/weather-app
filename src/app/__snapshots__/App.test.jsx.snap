// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <App/> component snapshot testing 1`] = `"<div class=\\"MuiBox-root css-1xmafg8\\"><div class=\\"MuiBox-root css-119w1of\\"><div class=\\"MuiBox-root css-0\\"><div aria-busy=\\"true\\" class=\\"\\"><svg width=\\"80\\" height=\\"80\\" viewBox=\\"0 0 120 30\\" xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"white\\" aria-label=\\"audio-loading\\"><circle cx=\\"15\\" cy=\\"15\\" r=\\"15\\"><animate attributeName=\\"r\\" from=\\"15\\" to=\\"15\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"15;9;15\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate><animate attributeName=\\"fillOpacity\\" from=\\"1\\" to=\\"1\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"1;.5;1\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate></circle><circle cx=\\"60\\" cy=\\"15\\" r=\\"9\\" attributeName=\\"fillOpacity\\" from=\\"1\\" to=\\"0.3\\"><animate attributeName=\\"r\\" from=\\"9\\" to=\\"9\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"9;15;9\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate><animate attributeName=\\"fillOpacity\\" from=\\"0.5\\" to=\\"0.5\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\".5;1;.5\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate></circle><circle cx=\\"105\\" cy=\\"15\\" r=\\"15\\"><animate attributeName=\\"r\\" from=\\"15\\" to=\\"15\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"15;9;15\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate><animate attributeName=\\"fillOpacity\\" from=\\"1\\" to=\\"1\\" begin=\\"0s\\" dur=\\"0.8s\\" values=\\"1;.5;1\\" calcMode=\\"linear\\" repeatCount=\\"indefinite\\"></animate></circle></svg></div><div class=\\"MuiBox-root css-0\\">In a moment . . .</div></div></div><div class=\\"MuiBox-root css-1y08cah\\"><lottie-player src=\\"https://assets3.lottiefiles.com/packages/lf20_xx4r8tdi.json\\" background=\\"transparent\\" speed=\\"1\\" style=\\"width: 100px; height: 100px;\\" loop=\\"true\\" autoplay=\\"true\\"></lottie-player></div></div>"`;
